tests/repository_simulator.py

```diff
@@ -313,9 +313,13 @@ def update_timestamp(self) -> None:
         if self.compute_metafile_hashes_length:
             hashes, length = self._compute_hashes_and_length(Snapshot.type)
 
-        self.timestamp.snapshot_meta = MetaFile(
-            self.snapshot.version, length, hashes
-        )
+        metafile_dict = {
+            "version": self.snapshot.version,
+            "length": length,
+            "hashes": hashes,
+        }
+
+        self.timestamp.snapshot_meta = MetaFile.from_dict(metafile_dict)
```

Comment:
from_dict() exists as a helper to deserialization, I think using them as constructors is a code smell.

Can you explain what the incompatibility is in the original code?

Reply:
### Reason for commit
Pyre throws the following warning:
`Expected int for 1st positional only parameter to call MetaFile.__init__ but got Optional[int].`

### Why use from_dict?
I used `MetaFile.from_dict()` thinking it is the convention as other tests use a similar approach.
https://github.com/theupdateframework/python-tuf/blob/c5ca38f0ae86cadad81e8adf99df1566cdee54da/tests/test_metadata_eq_.py#L167

Another approach would be to use a temporary var for `self.snapshot.version` and guard it with the if None clause.

Reply:
That's interesting... My first instinct is that this might be a bug in Pyre: `self.snapshot.version` should always be an int and this is how mypy (another static type checker) handles it. I can see how a static analyzer could be confused (Snapshot constructor does accept None as version) but the internal variable is never set to None.

Assuming my analysis is correct (and this is a bug in Pyre), I don't want more complex code like in this commit: I wouldn't object to changes to e.g. `Signed.__init__()` (where Snapshot.version is set) if they don't make the code harder to read and at the same time make life easier for static checkers.

You are correct that other tests do sometimes use "`from_dict()` as constructor" approach. I don't think it's good style and we should rather improve constructors if they are not good enough.

Reply:
There's a lot going in `Signed.__init__()` that I find interesting.

https://github.com/theupdateframework/python-tuf/blob/c5ca38f0ae86cadad81e8adf99df1566cdee54da/tuf/api/metadata.py#L495-L502

Do we need `Optional` type in `version` and `spec_versions` arguments in `__init__`? Both of them are immutable types.

````
    def __init__(
        self,
        version: int = 1,
        spec_version: str = ".".join(SPECIFICATION_VERSION),
        expires: Optional[datetime],
        unrecognized_fields: Optional[Dict[str, Any]],
    ):
````

Reply:
Yeah, it's a bit tricky. It's indeed possible that there is a design where Optional is not required for version, spec_version or even expires (immutability isn't required by python itself as you could set the default to None and then handle that in code) but it's not quite as simple as that as you probably found out... 

Basically:
* we want easy to use constructors that don't require spelling out all arguments
* we want to not write same code twice for constructor and the deserializing code paths -- which is why we currently call constructors from `from_dict()` helpers
* preferably this all happens without breaking current API

The current implementation mostly gives us these benefits **and** mypy can understand it...